{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "00910a35-ffbc-4497-bcc8-21f008c326fb",
   "metadata": {
    "tags": []
   },
   "source": [
    "# FinnGen - Data Engineering Notebook"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6fd8be8-14cc-4d80-a944-76c5235c078b",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Data Descriptions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1846d70-3c09-4b34-a05b-b72706e828fa",
   "metadata": {
    "tags": []
   },
   "source": [
    "### gwas"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8535f83a-6468-41bb-b3dc-6e678783c019",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "\n",
    "The `{endpoint}.gz` file has the following structure:\n",
    "\n",
    "| Column name   | Description                                                 |\n",
    "| ------------- | ----------------------------------------------------------- |\n",
    "| #chrom        | chromosome on build GRCh38 (1-23)                           |\n",
    "| pos           | position in base pairs on build GRCh38                       |\n",
    "| ref           | reference allele                                            |\n",
    "| alt           | alternative allele (effect allele)                           |\n",
    "| rsids         | variant identifier                                          |\n",
    "| nearest_genes | nearest gene(s) (comma separated) from variant               |\n",
    "| pval          | p-value from [source]                                        |\n",
    "| mlogp         | -log10(p-value)                                             |\n",
    "| beta          | effect size (log(OR) scale) estimated with [source]          |\n",
    "| sebeta        | standard error of effect size estimated with [source]        |\n",
    "| af_alt        | alternative (effect) allele frequency                        |\n",
    "| af_alt_cases  | alternative (effect) allele frequency among cases            |\n",
    "| af_alt_controls | alternative (effect) allele frequency among controls         |\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4fe3adfe-4cf2-4364-8436-13fbd76cb5b8",
   "metadata": {
    "tags": []
   },
   "source": [
    "### causal"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef7c3221-9d4e-480c-91fc-aff10d684f3b",
   "metadata": {},
   "source": [
    "Data taken from:\n",
    "\n",
    "[Functional characterization of T2D-associated SNP effects on baseline and ER stress-responsive β cell transcriptional activation](https://www.nature.com/articles/s41467-021-25514-6#MOESM8)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a7cea15-a49b-4ee2-87a5-bf625453328e",
   "metadata": {
    "tags": []
   },
   "source": [
    "### finemap"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10f3aafa-e866-463a-8cd1-a7416b1ff129",
   "metadata": {},
   "source": [
    "{endpoint}.SUSIE.snp.bgz` contains variant summaries with credible set information and has the following structure:\n",
    "\n",
    "| Column name    | Description                                                        |\n",
    "| -------------- | ------------------------------------------------------------------ |\n",
    "| trait          | endpoint name                                                      |\n",
    "| region         | chr:start-end                                                      |\n",
    "| v              | variant identifier                                                 |\n",
    "| rsid           | rs variant identifier                                              |\n",
    "| chromosome     | chromosome on build GRCh38 (1-22, X)                                |\n",
    "| position       | position in base pairs on build GRCh38                              |\n",
    "| allele1        | reference allele                                                   |\n",
    "| allele2        | alternative allele (effect allele)                                  |\n",
    "| maf            | minor allele frequency                                             |\n",
    "| beta           | effect size GWAS                                                   |\n",
    "| se             | standard error GWAS                                                |\n",
    "| p              | p-value GWAS                                                       |\n",
    "| mean           | posterior expectation of true effect size                           |\n",
    "| sd             | posterior standard deviation of true effect size                   |\n",
    "| prob           | posterior probability of association                                |\n",
    "| cs             | identifier of 95% credible set (-1 = variant is not part of credible set) |\n",
    "| lead_r2        | r2 value to a lead variant (the one with maximum PIP) in a credible set |\n",
    "| alphax         | posterior inclusion probability for the x-th single effect (x := 1..L where L is the number of single effects (causal variants) specified; default: L = 10) |"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "018acd43-12a7-41be-bf26-fc1d03dd7ef7",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "483d7bbb-33ef-4893-b03c-6173b7253544",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python version: 3.11.3 (tags/v3.11.3:f3909b8, Apr  4 2023, 23:49:59) [MSC v.1934 64 bit (AMD64)]\n",
      "Pandas version: 2.0.1\n",
      "NumPy version: 1.24.3\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import requests\n",
    "import time\n",
    "from concurrent.futures import ThreadPoolExecutor\n",
    "import myvariant\n",
    "\n",
    "\n",
    "\n",
    "print(\"Python version:\", sys.version)\n",
    "print(\"Pandas version:\", pd.__version__)\n",
    "print(\"NumPy version:\", np.__version__)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "445033a6-7000-4fb6-9c4d-cfe03f570b1c",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Import data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d4345459-b1ef-477c-8f06-c8f54c194af0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Read the 'finemap' file into a pandas DataFrame\n",
    "finemap = pd.read_csv('C:/Users/falty/Desktop/gwas-graph/FinnGen/data/finemapping_full_finngen_R9_T2D.SUSIE.snp.tsv', low_memory=False, sep='\\t')\n",
    "\n",
    "# Read the 'causal' file into a pandas DataFrame\n",
    "causal = pd.read_csv('C:/Users/falty/Desktop/gwas-graph/FinnGen/data/41467_2021_25514_MOESM8_ESM.csv', low_memory=False)\n",
    "\n",
    "# Read the 'gwas' file into a pandas DataFrame\n",
    "gwas = pd.read_csv('C:/Users/falty/Desktop/gwas-graph/FinnGen/data/summary_stats_finngen_R9_T2D.tsv', low_memory=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ffc11568-26b1-4283-b6b6-e1fd478f29e2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NaNs and missing values in 'gwas':\n",
      "#chrom                   0\n",
      "pos                      0\n",
      "ref                      0\n",
      "alt                      0\n",
      "rsids              1366396\n",
      "nearest_genes       727855\n",
      "pval                     0\n",
      "mlogp                    0\n",
      "beta                     0\n",
      "sebeta                   0\n",
      "af_alt                   0\n",
      "af_alt_cases             0\n",
      "af_alt_controls          0\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(\"NaNs and missing values in 'gwas':\")\n",
    "empty = gwas.isna().sum()\n",
    "print(empty)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3610b12e-d085-49a4-91e8-273fcb9dad32",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Explore data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e3742572-cf8b-48a5-a540-4b99d41324b1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>rsids</th>\n",
       "      <th>nearest_genes</th>\n",
       "      <th>pval</th>\n",
       "      <th>mlogp</th>\n",
       "      <th>beta</th>\n",
       "      <th>sebeta</th>\n",
       "      <th>af_alt</th>\n",
       "      <th>af_alt_cases</th>\n",
       "      <th>af_alt_controls</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>13668</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>rs2691328</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>0.944365</td>\n",
       "      <td>0.024860</td>\n",
       "      <td>-0.005926</td>\n",
       "      <td>0.084918</td>\n",
       "      <td>0.005842</td>\n",
       "      <td>0.005729</td>\n",
       "      <td>0.005863</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>14773</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>rs878915777</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>0.844305</td>\n",
       "      <td>0.073501</td>\n",
       "      <td>0.010088</td>\n",
       "      <td>0.051369</td>\n",
       "      <td>0.013495</td>\n",
       "      <td>0.013547</td>\n",
       "      <td>0.013485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>15585</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>rs533630043</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>0.841908</td>\n",
       "      <td>0.074735</td>\n",
       "      <td>0.031464</td>\n",
       "      <td>0.157751</td>\n",
       "      <td>0.001113</td>\n",
       "      <td>0.001125</td>\n",
       "      <td>0.001110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>16549</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>rs1262014613</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>0.343308</td>\n",
       "      <td>0.464316</td>\n",
       "      <td>0.241377</td>\n",
       "      <td>0.254711</td>\n",
       "      <td>0.000561</td>\n",
       "      <td>0.000620</td>\n",
       "      <td>0.000550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>16567</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>rs1194064194</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>0.129883</td>\n",
       "      <td>0.886447</td>\n",
       "      <td>0.130736</td>\n",
       "      <td>0.086319</td>\n",
       "      <td>0.004170</td>\n",
       "      <td>0.004250</td>\n",
       "      <td>0.004154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20170001</th>\n",
       "      <td>23</td>\n",
       "      <td>155697920</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.027115</td>\n",
       "      <td>1.566790</td>\n",
       "      <td>-0.013475</td>\n",
       "      <td>0.006097</td>\n",
       "      <td>0.290961</td>\n",
       "      <td>0.286054</td>\n",
       "      <td>0.291879</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20170002</th>\n",
       "      <td>23</td>\n",
       "      <td>155698443</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.178417</td>\n",
       "      <td>0.748564</td>\n",
       "      <td>-0.069907</td>\n",
       "      <td>0.051951</td>\n",
       "      <td>0.003259</td>\n",
       "      <td>0.003022</td>\n",
       "      <td>0.003304</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20170003</th>\n",
       "      <td>23</td>\n",
       "      <td>155698490</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.279640</td>\n",
       "      <td>0.553400</td>\n",
       "      <td>-0.020245</td>\n",
       "      <td>0.018725</td>\n",
       "      <td>0.024406</td>\n",
       "      <td>0.024312</td>\n",
       "      <td>0.024423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20170004</th>\n",
       "      <td>23</td>\n",
       "      <td>155699751</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.078864</td>\n",
       "      <td>1.103120</td>\n",
       "      <td>-0.011284</td>\n",
       "      <td>0.006421</td>\n",
       "      <td>0.244829</td>\n",
       "      <td>0.241257</td>\n",
       "      <td>0.245498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20170005</th>\n",
       "      <td>23</td>\n",
       "      <td>155700291</td>\n",
       "      <td>CAA</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.071590</td>\n",
       "      <td>1.145150</td>\n",
       "      <td>-0.011563</td>\n",
       "      <td>0.006418</td>\n",
       "      <td>0.245328</td>\n",
       "      <td>0.241724</td>\n",
       "      <td>0.246003</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20170006 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          #chrom        pos  ref alt         rsids nearest_genes      pval   \n",
       "0              1      13668    G   A     rs2691328         OR4F5  0.944365  \\\n",
       "1              1      14773    C   T   rs878915777         OR4F5  0.844305   \n",
       "2              1      15585    G   A   rs533630043         OR4F5  0.841908   \n",
       "3              1      16549    T   C  rs1262014613         OR4F5  0.343308   \n",
       "4              1      16567    G   C  rs1194064194         OR4F5  0.129883   \n",
       "...          ...        ...  ...  ..           ...           ...       ...   \n",
       "20170001      23  155697920    G   A           NaN           NaN  0.027115   \n",
       "20170002      23  155698443    C   A           NaN           NaN  0.178417   \n",
       "20170003      23  155698490    C   T           NaN           NaN  0.279640   \n",
       "20170004      23  155699751    C   T           NaN           NaN  0.078864   \n",
       "20170005      23  155700291  CAA   C           NaN           NaN  0.071590   \n",
       "\n",
       "             mlogp      beta    sebeta    af_alt  af_alt_cases   \n",
       "0         0.024860 -0.005926  0.084918  0.005842      0.005729  \\\n",
       "1         0.073501  0.010088  0.051369  0.013495      0.013547   \n",
       "2         0.074735  0.031464  0.157751  0.001113      0.001125   \n",
       "3         0.464316  0.241377  0.254711  0.000561      0.000620   \n",
       "4         0.886447  0.130736  0.086319  0.004170      0.004250   \n",
       "...            ...       ...       ...       ...           ...   \n",
       "20170001  1.566790 -0.013475  0.006097  0.290961      0.286054   \n",
       "20170002  0.748564 -0.069907  0.051951  0.003259      0.003022   \n",
       "20170003  0.553400 -0.020245  0.018725  0.024406      0.024312   \n",
       "20170004  1.103120 -0.011284  0.006421  0.244829      0.241257   \n",
       "20170005  1.145150 -0.011563  0.006418  0.245328      0.241724   \n",
       "\n",
       "          af_alt_controls  \n",
       "0                0.005863  \n",
       "1                0.013485  \n",
       "2                0.001110  \n",
       "3                0.000550  \n",
       "4                0.004154  \n",
       "...                   ...  \n",
       "20170001         0.291879  \n",
       "20170002         0.003304  \n",
       "20170003         0.024423  \n",
       "20170004         0.245498  \n",
       "20170005         0.246003  \n",
       "\n",
       "[20170006 rows x 13 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8c5681a2-85f4-4fd3-a3b9-de8824651d5c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Trait</th>\n",
       "      <th>IndexSNP</th>\n",
       "      <th>SNP</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>LOCUS</th>\n",
       "      <th>REF_ALLELE</th>\n",
       "      <th>ALT_ALLELE</th>\n",
       "      <th>MAF_AFR</th>\n",
       "      <th>MAF_AMR</th>\n",
       "      <th>...</th>\n",
       "      <th>REF_HIGHER_Under_Erstress</th>\n",
       "      <th>ALT_HIGHER_Under_Erstress</th>\n",
       "      <th>ATAC_Peak</th>\n",
       "      <th>PDX1</th>\n",
       "      <th>FOXA2</th>\n",
       "      <th>H2A.Z</th>\n",
       "      <th>H3K27ac</th>\n",
       "      <th>MAFB</th>\n",
       "      <th>NKX6.1</th>\n",
       "      <th>CTCF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs2820446</td>\n",
       "      <td>rs4846567</td>\n",
       "      <td>1</td>\n",
       "      <td>219750717</td>\n",
       "      <td>LYPAL1</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.41</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs35072907</td>\n",
       "      <td>rs72904726</td>\n",
       "      <td>1</td>\n",
       "      <td>51339332</td>\n",
       "      <td>FAF1</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.08</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs1861612</td>\n",
       "      <td>rs10197480</td>\n",
       "      <td>2</td>\n",
       "      <td>230521253</td>\n",
       "      <td>DNER</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.53</td>\n",
       "      <td>0.54</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs34669198</td>\n",
       "      <td>rs13026123</td>\n",
       "      <td>2</td>\n",
       "      <td>45654050</td>\n",
       "      <td>SRBD1</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.09</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs3923113</td>\n",
       "      <td>rs6713419</td>\n",
       "      <td>2</td>\n",
       "      <td>165508300</td>\n",
       "      <td>GRB14,COBLL1</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.76</td>\n",
       "      <td>0.29</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs6712932</td>\n",
       "      <td>rs10210149</td>\n",
       "      <td>2</td>\n",
       "      <td>105839324</td>\n",
       "      <td>GPR45, LINC01918</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.27</td>\n",
       "      <td>0.29</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs6723108</td>\n",
       "      <td>rs6723108</td>\n",
       "      <td>2</td>\n",
       "      <td>135479980</td>\n",
       "      <td>TMEM163/RAB3GAP1/ACMSD</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.81</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs7578326</td>\n",
       "      <td>rs2943656</td>\n",
       "      <td>2</td>\n",
       "      <td>227121918</td>\n",
       "      <td>IRS1</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.73</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs7578597</td>\n",
       "      <td>rs13405776</td>\n",
       "      <td>2</td>\n",
       "      <td>43739121</td>\n",
       "      <td>THADA</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.34</td>\n",
       "      <td>0.07</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs9309245</td>\n",
       "      <td>rs10194328</td>\n",
       "      <td>2</td>\n",
       "      <td>53398669</td>\n",
       "      <td>ASB3</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.57</td>\n",
       "      <td>0.34</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs11717195</td>\n",
       "      <td>rs56371916</td>\n",
       "      <td>3</td>\n",
       "      <td>123095543</td>\n",
       "      <td>ADCY5</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.28</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs16861329</td>\n",
       "      <td>rs9798820</td>\n",
       "      <td>3</td>\n",
       "      <td>186676482</td>\n",
       "      <td>ST6GAL1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.22</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs2688419</td>\n",
       "      <td>rs2688419</td>\n",
       "      <td>3</td>\n",
       "      <td>23077761</td>\n",
       "      <td>RPL24P7,SALL4P5</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.31</td>\n",
       "      <td>0.35</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs4607103</td>\n",
       "      <td>rs147257399</td>\n",
       "      <td>3</td>\n",
       "      <td>64728255</td>\n",
       "      <td>ADAMTS9</td>\n",
       "      <td>CAA</td>\n",
       "      <td>C</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs73230612</td>\n",
       "      <td>rs9879646</td>\n",
       "      <td>3</td>\n",
       "      <td>114894528</td>\n",
       "      <td>ZBTB20</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.15</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs7630877</td>\n",
       "      <td>rs17748864</td>\n",
       "      <td>3</td>\n",
       "      <td>179649511</td>\n",
       "      <td>PEX5L</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.24</td>\n",
       "      <td>0.27</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs831571</td>\n",
       "      <td>rs76647135</td>\n",
       "      <td>3</td>\n",
       "      <td>64011584</td>\n",
       "      <td>PSMD6</td>\n",
       "      <td>AC</td>\n",
       "      <td>A</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.13</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs1048886</td>\n",
       "      <td>rs2881632</td>\n",
       "      <td>6</td>\n",
       "      <td>71296609</td>\n",
       "      <td>SDHAF4</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.40</td>\n",
       "      <td>0.22</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs1535500</td>\n",
       "      <td>rs12663159</td>\n",
       "      <td>6</td>\n",
       "      <td>39282277</td>\n",
       "      <td>KCNK16</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.86</td>\n",
       "      <td>0.49</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs9472138</td>\n",
       "      <td>rs12189774</td>\n",
       "      <td>6</td>\n",
       "      <td>43816126</td>\n",
       "      <td>VEGFA (Ahluwalia et al 2017; implication of CHOP)</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.21</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs7636</td>\n",
       "      <td>rs57058793</td>\n",
       "      <td>7</td>\n",
       "      <td>100498220</td>\n",
       "      <td>ACHE</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.29</td>\n",
       "      <td>0.08</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs6467136</td>\n",
       "      <td>rs7783500</td>\n",
       "      <td>7</td>\n",
       "      <td>127191947</td>\n",
       "      <td>AC000124.1, GCC1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.54</td>\n",
       "      <td>0.40</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs1561927</td>\n",
       "      <td>rs7824269</td>\n",
       "      <td>8</td>\n",
       "      <td>129552344</td>\n",
       "      <td>LINC00824</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.63</td>\n",
       "      <td>0.30</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs10741243</td>\n",
       "      <td>rs4341449</td>\n",
       "      <td>10</td>\n",
       "      <td>132949735</td>\n",
       "      <td>TCERG1L</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.84</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs35668226</td>\n",
       "      <td>rs11199771</td>\n",
       "      <td>10</td>\n",
       "      <td>122868214</td>\n",
       "      <td>AC010998.3,RPL19P16</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.15</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs7087591</td>\n",
       "      <td>rs2901609</td>\n",
       "      <td>10</td>\n",
       "      <td>94475111</td>\n",
       "      <td>Y_RNA, EXOC6</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.35</td>\n",
       "      <td>0.37</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs7903146</td>\n",
       "      <td>rs7903146</td>\n",
       "      <td>10</td>\n",
       "      <td>114758349</td>\n",
       "      <td>TCF7L2</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.28</td>\n",
       "      <td>0.25</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs11564710</td>\n",
       "      <td>rs11564710</td>\n",
       "      <td>11</td>\n",
       "      <td>2200329</td>\n",
       "      <td>TH</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.45</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs1387153</td>\n",
       "      <td>rs11020114</td>\n",
       "      <td>11</td>\n",
       "      <td>92682604</td>\n",
       "      <td>MTNR1B</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.43</td>\n",
       "      <td>0.29</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs1552224</td>\n",
       "      <td>rs11603349</td>\n",
       "      <td>11</td>\n",
       "      <td>72460694</td>\n",
       "      <td>STARD10</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.09</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs2722769</td>\n",
       "      <td>rs2655100</td>\n",
       "      <td>11</td>\n",
       "      <td>11228528</td>\n",
       "      <td>LINC02752,MTND5P21</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.31</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs3842770</td>\n",
       "      <td>rs115420895</td>\n",
       "      <td>11</td>\n",
       "      <td>2197121</td>\n",
       "      <td>AC132217.2,INS-IGF2</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.01</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs5215</td>\n",
       "      <td>rs1002226</td>\n",
       "      <td>11</td>\n",
       "      <td>17405617</td>\n",
       "      <td>KCNJ11</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.91</td>\n",
       "      <td>0.67</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs9300039</td>\n",
       "      <td>rs12788548</td>\n",
       "      <td>11</td>\n",
       "      <td>41906381</td>\n",
       "      <td>LINC02745, AC023442.2</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.13</td>\n",
       "      <td>0.06</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs7957197</td>\n",
       "      <td>rs7957197</td>\n",
       "      <td>12</td>\n",
       "      <td>121460686</td>\n",
       "      <td>HNF1A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.17</td>\n",
       "      <td>0.15</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs7978895</td>\n",
       "      <td>rs4565964</td>\n",
       "      <td>12</td>\n",
       "      <td>43575002</td>\n",
       "      <td>ADAMTS20, LINC02461</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.60</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs10507349</td>\n",
       "      <td>rs4630391</td>\n",
       "      <td>13</td>\n",
       "      <td>26777824</td>\n",
       "      <td>RNF6</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.36</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs1359790</td>\n",
       "      <td>rs12428731</td>\n",
       "      <td>13</td>\n",
       "      <td>80709595</td>\n",
       "      <td>SPRY2</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.30</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs79548680</td>\n",
       "      <td>rs79548680</td>\n",
       "      <td>15</td>\n",
       "      <td>91505779</td>\n",
       "      <td>RCCD1</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.14</td>\n",
       "      <td>0.25</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs17797882</td>\n",
       "      <td>rs76544785</td>\n",
       "      <td>16</td>\n",
       "      <td>79404197</td>\n",
       "      <td>WWOX</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.06</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs275856</td>\n",
       "      <td>rs275856</td>\n",
       "      <td>18</td>\n",
       "      <td>21839152</td>\n",
       "      <td>OSBPL1A</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.06</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs6506284</td>\n",
       "      <td>rs6506284</td>\n",
       "      <td>18</td>\n",
       "      <td>5030023</td>\n",
       "      <td>AP005203.1, LINC01892</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.02</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>T2D</td>\n",
       "      <td>rs4898364</td>\n",
       "      <td>rs4898364</td>\n",
       "      <td>X</td>\n",
       "      <td>152900485</td>\n",
       "      <td>DUSP9</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.17</td>\n",
       "      <td>0.49</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>43 rows × 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Trait    IndexSNP          SNP CHR   POSITION   \n",
       "0    T2D   rs2820446    rs4846567   1  219750717  \\\n",
       "1    T2D  rs35072907   rs72904726   1   51339332   \n",
       "2    T2D   rs1861612   rs10197480   2  230521253   \n",
       "3    T2D  rs34669198   rs13026123   2   45654050   \n",
       "4    T2D   rs3923113    rs6713419   2  165508300   \n",
       "5    T2D   rs6712932   rs10210149   2  105839324   \n",
       "6    T2D   rs6723108    rs6723108   2  135479980   \n",
       "7    T2D   rs7578326    rs2943656   2  227121918   \n",
       "8    T2D   rs7578597   rs13405776   2   43739121   \n",
       "9    T2D   rs9309245   rs10194328   2   53398669   \n",
       "10   T2D  rs11717195   rs56371916   3  123095543   \n",
       "11   T2D  rs16861329    rs9798820   3  186676482   \n",
       "12   T2D   rs2688419    rs2688419   3   23077761   \n",
       "13   T2D   rs4607103  rs147257399   3   64728255   \n",
       "14   T2D  rs73230612    rs9879646   3  114894528   \n",
       "15   T2D   rs7630877   rs17748864   3  179649511   \n",
       "16   T2D    rs831571   rs76647135   3   64011584   \n",
       "17   T2D   rs1048886    rs2881632   6   71296609   \n",
       "18   T2D   rs1535500   rs12663159   6   39282277   \n",
       "19   T2D   rs9472138   rs12189774   6   43816126   \n",
       "20   T2D      rs7636   rs57058793   7  100498220   \n",
       "21   T2D   rs6467136    rs7783500   7  127191947   \n",
       "22   T2D   rs1561927    rs7824269   8  129552344   \n",
       "23   T2D  rs10741243    rs4341449  10  132949735   \n",
       "24   T2D  rs35668226   rs11199771  10  122868214   \n",
       "25   T2D   rs7087591    rs2901609  10   94475111   \n",
       "26   T2D   rs7903146    rs7903146  10  114758349   \n",
       "27   T2D  rs11564710   rs11564710  11    2200329   \n",
       "28   T2D   rs1387153   rs11020114  11   92682604   \n",
       "29   T2D   rs1552224   rs11603349  11   72460694   \n",
       "30   T2D   rs2722769    rs2655100  11   11228528   \n",
       "31   T2D   rs3842770  rs115420895  11    2197121   \n",
       "32   T2D      rs5215    rs1002226  11   17405617   \n",
       "33   T2D   rs9300039   rs12788548  11   41906381   \n",
       "34   T2D   rs7957197    rs7957197  12  121460686   \n",
       "35   T2D   rs7978895    rs4565964  12   43575002   \n",
       "36   T2D  rs10507349    rs4630391  13   26777824   \n",
       "37   T2D   rs1359790   rs12428731  13   80709595   \n",
       "38   T2D  rs79548680   rs79548680  15   91505779   \n",
       "39   T2D  rs17797882   rs76544785  16   79404197   \n",
       "40   T2D    rs275856     rs275856  18   21839152   \n",
       "41   T2D   rs6506284    rs6506284  18    5030023   \n",
       "42   T2D   rs4898364    rs4898364   X  152900485   \n",
       "\n",
       "                                                LOCUS REF_ALLELE ALT_ALLELE   \n",
       "0                                              LYPAL1          G          T  \\\n",
       "1                                                FAF1          T          C   \n",
       "2                                                DNER          C          T   \n",
       "3                                               SRBD1          T          A   \n",
       "4                                        GRB14,COBLL1          T          C   \n",
       "5                                    GPR45, LINC01918          C          T   \n",
       "6                              TMEM163/RAB3GAP1/ACMSD          G          T   \n",
       "7                                                IRS1          A          G   \n",
       "8                                               THADA          C          T   \n",
       "9                                                ASB3          G          T   \n",
       "10                                              ADCY5          T          C   \n",
       "11                                            ST6GAL1          G          A   \n",
       "12                                    RPL24P7,SALL4P5          T          C   \n",
       "13                                            ADAMTS9        CAA          C   \n",
       "14                                             ZBTB20          G          A   \n",
       "15                                              PEX5L          C          T   \n",
       "16                                              PSMD6         AC          A   \n",
       "17                                             SDHAF4          C          A   \n",
       "18                                             KCNK16          C          A   \n",
       "19  VEGFA (Ahluwalia et al 2017; implication of CHOP)          G          A   \n",
       "20                                               ACHE          C          T   \n",
       "21                                   AC000124.1, GCC1          G          A   \n",
       "22                                          LINC00824          C          T   \n",
       "23                                            TCERG1L          C          T   \n",
       "24                                AC010998.3,RPL19P16          A          G   \n",
       "25                                       Y_RNA, EXOC6          G          A   \n",
       "26                                             TCF7L2          C          T   \n",
       "27                                                 TH          A          C   \n",
       "28                                             MTNR1B          T          C   \n",
       "29                                            STARD10          T          C   \n",
       "30                                 LINC02752,MTND5P21          A          G   \n",
       "31                                AC132217.2,INS-IGF2          G          A   \n",
       "32                                             KCNJ11          C          T   \n",
       "33                              LINC02745, AC023442.2          C          T   \n",
       "34                                              HNF1A          T          A   \n",
       "35                                ADAMTS20, LINC02461          A          G   \n",
       "36                                               RNF6          C          T   \n",
       "37                                              SPRY2          G          T   \n",
       "38                                              RCCD1          G          C   \n",
       "39                                               WWOX          G          C   \n",
       "40                                            OSBPL1A          A          T   \n",
       "41                              AP005203.1, LINC01892          C          T   \n",
       "42                                              DUSP9          C          T   \n",
       "\n",
       "    MAF_AFR  MAF_AMR  ...  REF_HIGHER_Under_Erstress   \n",
       "0      0.05     0.41  ...                        0.0  \\\n",
       "1      0.02     0.08  ...                        0.0   \n",
       "2      0.53     0.54  ...                        0.0   \n",
       "3      0.09     0.09  ...                        1.0   \n",
       "4      0.76     0.29  ...                        0.0   \n",
       "5      0.27     0.29  ...                        0.0   \n",
       "6      0.99     0.81  ...                        0.0   \n",
       "7      0.60     0.73  ...                        0.0   \n",
       "8      0.34     0.07  ...                        0.0   \n",
       "9      0.57     0.34  ...                        0.0   \n",
       "10     0.09     0.28  ...                        0.0   \n",
       "11     0.05     0.22  ...                        1.0   \n",
       "12     0.31     0.35  ...                        0.0   \n",
       "13    -1.00    -1.00  ...                        1.0   \n",
       "14     0.12     0.15  ...                        1.0   \n",
       "15     0.24     0.27  ...                        0.0   \n",
       "16     0.19     0.13  ...                        1.0   \n",
       "17     0.40     0.22  ...                        1.0   \n",
       "18     0.86     0.49  ...                        0.0   \n",
       "19     0.19     0.21  ...                       -1.0   \n",
       "20     0.29     0.08  ...                        0.0   \n",
       "21     0.54     0.40  ...                        0.0   \n",
       "22     0.63     0.30  ...                       -1.0   \n",
       "23     0.56     0.84  ...                        1.0   \n",
       "24     0.05     0.15  ...                        0.0   \n",
       "25     0.35     0.37  ...                        0.0   \n",
       "26     0.28     0.25  ...                        0.0   \n",
       "27     0.11     0.45  ...                       -1.0   \n",
       "28     0.43     0.29  ...                        0.0   \n",
       "29     0.03     0.09  ...                        1.0   \n",
       "30     0.03     0.31  ...                        1.0   \n",
       "31     0.06     0.01  ...                        0.0   \n",
       "32     0.91     0.67  ...                        0.0   \n",
       "33     0.13     0.06  ...                        NaN   \n",
       "34     0.17     0.15  ...                        0.0   \n",
       "35     0.48     0.60  ...                        0.0   \n",
       "36     0.08     0.36  ...                        0.0   \n",
       "37     0.19     0.30  ...                        0.0   \n",
       "38     0.14     0.25  ...                        0.0   \n",
       "39     0.09     0.06  ...                        0.0   \n",
       "40     0.22     0.06  ...                        1.0   \n",
       "41     0.22     0.02  ...                        0.0   \n",
       "42     0.17     0.49  ...                        0.0   \n",
       "\n",
       "    ALT_HIGHER_Under_Erstress ATAC_Peak PDX1 FOXA2  H2A.Z  H3K27ac  MAFB   \n",
       "0                          -1         1    0     0      0        0     0  \\\n",
       "1                           0         1    0     0      0        0     0   \n",
       "2                           0         0    0     0      0        0     0   \n",
       "3                           0         0    0     0      0        0     0   \n",
       "4                           0         0    0     0      0        0     0   \n",
       "5                           0         1    0     0      0        0     0   \n",
       "6                           0         0    0     0      0        0     0   \n",
       "7                           0         1    0     0      1        0     0   \n",
       "8                           0         0    0     0      0        0     0   \n",
       "9                           0         0    0     0      0        0     0   \n",
       "10                          0         0    0     0      0        0     0   \n",
       "11                          1         0    0     0      0        0     0   \n",
       "12                          0         0    0     0      0        0     0   \n",
       "13                          1         0    0     0      0        0     0   \n",
       "14                          1         0    0     0      0        0     0   \n",
       "15                          1         0    0     0      0        0     0   \n",
       "16                          1         0    0     0      0        0     0   \n",
       "17                          1         0    0     0      0        0     0   \n",
       "18                          0         1    0     0      0        0     0   \n",
       "19                         -1         1    1     0      0        1     1   \n",
       "20                          1         0    0     0      0        0     0   \n",
       "21                          0         0    0     0      0        0     0   \n",
       "22                         -1         1    0     0      0        1     0   \n",
       "23                          0         0    0     0      0        0     0   \n",
       "24                          0         0    0     0      0        0     0   \n",
       "25                          0         0    0     0      0        0     0   \n",
       "26                         -1         1    1     1      1        1     1   \n",
       "27                          0         1    1     1      0        0     0   \n",
       "28                          0         0    0     0      0        0     0   \n",
       "29                          1         0    0     0      0        0     0   \n",
       "30                          0         0    0     0      0        0     0   \n",
       "31                         -1         1    1     0      0        0     1   \n",
       "32                          0         1    0     0      0        0     0   \n",
       "33                          0         0    0     0      0        0     0   \n",
       "34                          0         0    0     0      0        0     0   \n",
       "35                          0         0    0     0      0        0     0   \n",
       "36                          1         1    0     0      1        1     0   \n",
       "37                          1         0    0     0      0        0     0   \n",
       "38                          0         0    0     0      0        0     0   \n",
       "39                          0         0    0     0      0        0     0   \n",
       "40                          1         0    0     0      0        0     0   \n",
       "41                          0         0    0     0      0        0     0   \n",
       "42                          0         0    0     0      0        0     0   \n",
       "\n",
       "    NKX6.1  CTCF  \n",
       "0        0     0  \n",
       "1        0     1  \n",
       "2        0     0  \n",
       "3        0     0  \n",
       "4        0     0  \n",
       "5        0     0  \n",
       "6        0     0  \n",
       "7        0     0  \n",
       "8        0     0  \n",
       "9        0     0  \n",
       "10       0     0  \n",
       "11       0     0  \n",
       "12       0     0  \n",
       "13       0     0  \n",
       "14       0     0  \n",
       "15       0     0  \n",
       "16       0     0  \n",
       "17       0     0  \n",
       "18       0     0  \n",
       "19       0     0  \n",
       "20       0     0  \n",
       "21       0     0  \n",
       "22       0     0  \n",
       "23       0     0  \n",
       "24       0     0  \n",
       "25       0     0  \n",
       "26       1     0  \n",
       "27       1     0  \n",
       "28       0     0  \n",
       "29       0     0  \n",
       "30       0     0  \n",
       "31       0     0  \n",
       "32       0     1  \n",
       "33       0     0  \n",
       "34       0     0  \n",
       "35       0     0  \n",
       "36       0     0  \n",
       "37       0     0  \n",
       "38       0     0  \n",
       "39       0     0  \n",
       "40       0     0  \n",
       "41       0     0  \n",
       "42       0     0  \n",
       "\n",
       "[43 rows x 43 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "causal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "dc659323-8943-4e41-9f41-5cd0c8e2c17e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>trait</th>\n",
       "      <th>region</th>\n",
       "      <th>v</th>\n",
       "      <th>rsid</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>allele1</th>\n",
       "      <th>allele2</th>\n",
       "      <th>maf</th>\n",
       "      <th>beta</th>\n",
       "      <th>...</th>\n",
       "      <th>lbf_variable1</th>\n",
       "      <th>lbf_variable2</th>\n",
       "      <th>lbf_variable3</th>\n",
       "      <th>lbf_variable4</th>\n",
       "      <th>lbf_variable5</th>\n",
       "      <th>lbf_variable6</th>\n",
       "      <th>lbf_variable7</th>\n",
       "      <th>lbf_variable8</th>\n",
       "      <th>lbf_variable9</th>\n",
       "      <th>lbf_variable10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T2D</td>\n",
       "      <td>chr1:18908743-21908743</td>\n",
       "      <td>1:18908743:G:A</td>\n",
       "      <td>chr1_18908743_G_A</td>\n",
       "      <td>chr1</td>\n",
       "      <td>18908743</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.000438</td>\n",
       "      <td>-0.143377</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.448766</td>\n",
       "      <td>-0.536436</td>\n",
       "      <td>-0.533681</td>\n",
       "      <td>-0.531823</td>\n",
       "      <td>-0.531294</td>\n",
       "      <td>-0.532036</td>\n",
       "      <td>-0.533612</td>\n",
       "      <td>-0.535413</td>\n",
       "      <td>-0.536846</td>\n",
       "      <td>-0.537469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T2D</td>\n",
       "      <td>chr1:18908743-21908743</td>\n",
       "      <td>1:18909023:T:C</td>\n",
       "      <td>chr1_18909023_T_C</td>\n",
       "      <td>chr1</td>\n",
       "      <td>18909023</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.198418</td>\n",
       "      <td>-0.008341</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.325844</td>\n",
       "      <td>-0.352317</td>\n",
       "      <td>-0.350020</td>\n",
       "      <td>-0.348464</td>\n",
       "      <td>-0.348012</td>\n",
       "      <td>-0.348617</td>\n",
       "      <td>-0.349921</td>\n",
       "      <td>-0.351417</td>\n",
       "      <td>-0.352614</td>\n",
       "      <td>-0.353139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T2D</td>\n",
       "      <td>chr1:18908743-21908743</td>\n",
       "      <td>1:18909112:G:A</td>\n",
       "      <td>chr1_18909112_G_A</td>\n",
       "      <td>chr1</td>\n",
       "      <td>18909112</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.159798</td>\n",
       "      <td>-0.006078</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.561368</td>\n",
       "      <td>-0.529649</td>\n",
       "      <td>-0.526932</td>\n",
       "      <td>-0.525097</td>\n",
       "      <td>-0.524573</td>\n",
       "      <td>-0.525301</td>\n",
       "      <td>-0.526853</td>\n",
       "      <td>-0.528628</td>\n",
       "      <td>-0.530041</td>\n",
       "      <td>-0.530656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T2D</td>\n",
       "      <td>chr1:18908743-21908743</td>\n",
       "      <td>1:18909164:T:C</td>\n",
       "      <td>chr1_18909164_T_C</td>\n",
       "      <td>chr1</td>\n",
       "      <td>18909164</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.168400</td>\n",
       "      <td>-0.004160</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.657574</td>\n",
       "      <td>-0.652146</td>\n",
       "      <td>-0.649198</td>\n",
       "      <td>-0.647204</td>\n",
       "      <td>-0.646632</td>\n",
       "      <td>-0.647420</td>\n",
       "      <td>-0.649103</td>\n",
       "      <td>-0.651028</td>\n",
       "      <td>-0.652563</td>\n",
       "      <td>-0.653231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T2D</td>\n",
       "      <td>chr1:18908743-21908743</td>\n",
       "      <td>1:18909192:C:CTACG</td>\n",
       "      <td>chr1_18909192_C_CTACG</td>\n",
       "      <td>chr1</td>\n",
       "      <td>18909192</td>\n",
       "      <td>C</td>\n",
       "      <td>CTACG</td>\n",
       "      <td>0.000248</td>\n",
       "      <td>0.149161</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.565444</td>\n",
       "      <td>-0.628124</td>\n",
       "      <td>-0.625089</td>\n",
       "      <td>-0.623034</td>\n",
       "      <td>-0.622444</td>\n",
       "      <td>-0.623254</td>\n",
       "      <td>-0.624986</td>\n",
       "      <td>-0.626968</td>\n",
       "      <td>-0.628549</td>\n",
       "      <td>-0.629239</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 66 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  trait                  region                   v                   rsid   \n",
       "0   T2D  chr1:18908743-21908743      1:18908743:G:A      chr1_18908743_G_A  \\\n",
       "1   T2D  chr1:18908743-21908743      1:18909023:T:C      chr1_18909023_T_C   \n",
       "2   T2D  chr1:18908743-21908743      1:18909112:G:A      chr1_18909112_G_A   \n",
       "3   T2D  chr1:18908743-21908743      1:18909164:T:C      chr1_18909164_T_C   \n",
       "4   T2D  chr1:18908743-21908743  1:18909192:C:CTACG  chr1_18909192_C_CTACG   \n",
       "\n",
       "  chromosome  position allele1 allele2       maf      beta  ...   \n",
       "0       chr1  18908743       G       A  0.000438 -0.143377  ...  \\\n",
       "1       chr1  18909023       T       C  0.198418 -0.008341  ...   \n",
       "2       chr1  18909112       G       A  0.159798 -0.006078  ...   \n",
       "3       chr1  18909164       T       C  0.168400 -0.004160  ...   \n",
       "4       chr1  18909192       C   CTACG  0.000248  0.149161  ...   \n",
       "\n",
       "   lbf_variable1  lbf_variable2  lbf_variable3  lbf_variable4  lbf_variable5   \n",
       "0      -1.448766      -0.536436      -0.533681      -0.531823      -0.531294  \\\n",
       "1      -1.325844      -0.352317      -0.350020      -0.348464      -0.348012   \n",
       "2      -1.561368      -0.529649      -0.526932      -0.525097      -0.524573   \n",
       "3      -1.657574      -0.652146      -0.649198      -0.647204      -0.646632   \n",
       "4      -1.565444      -0.628124      -0.625089      -0.623034      -0.622444   \n",
       "\n",
       "   lbf_variable6  lbf_variable7  lbf_variable8  lbf_variable9  lbf_variable10  \n",
       "0      -0.532036      -0.533612      -0.535413      -0.536846       -0.537469  \n",
       "1      -0.348617      -0.349921      -0.351417      -0.352614       -0.353139  \n",
       "2      -0.525301      -0.526853      -0.528628      -0.530041       -0.530656  \n",
       "3      -0.647420      -0.649103      -0.651028      -0.652563       -0.653231  \n",
       "4      -0.623254      -0.624986      -0.626968      -0.628549       -0.629239  \n",
       "\n",
       "[5 rows x 66 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "finemap.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "95020fad-f610-4ae3-876a-ebcaef2b2985",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def explore_dataframe(dataframe, dataframe_name):\n",
    "    print(\"=== DataFrame Exploration: {} ===\".format(dataframe_name))\n",
    "    print(\"Number of Rows: {}\".format(dataframe.shape[0]))\n",
    "    print(\"Number of Columns: {}\".format(dataframe.shape[1]))\n",
    "    print(\"Column Names: {}\".format(\", \".join(dataframe.columns)))\n",
    "    print(\"\\nData Types of Columns:\")\n",
    "    print(dataframe.dtypes)\n",
    "    print(\"\\nNull Value Counts:\")\n",
    "    print(dataframe.isnull().sum())\n",
    "    print(\"\\nSummary Statistics:\")\n",
    "    print(dataframe.describe())\n",
    "    print(\"=== End of DataFrame Exploration: {} ===\\n\".format(dataframe_name))\n",
    "    \n",
    "#explore_dataframe(gwas, \"gwas\")\n",
    "#explore_dataframe(causal, \"causal\")\n",
    "#explore_dataframe(finemap, \"finemap\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c46b7243-adf9-42ac-9026-c5ac411159b6",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Data manipulation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8573bef-1cda-4e57-b3fa-f2b40333620a",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Adjust `chromosome` in `finemap`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "825d60e9-ce09-43d2-8b4d-e5f47a942e90",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Extract number from 'chromosome' and replace 'X' with '23'\n",
    "finemap['chromosome'] = finemap['chromosome'].str.extract('(\\d+|X)', expand=False).replace('X', '23')\n",
    "\n",
    "# Convert 'chromosome' column to 'int64'\n",
    "finemap['chromosome'] = finemap['chromosome'].astype('int64')\n",
    "\n",
    "# Assertions to verify the data manipulations\n",
    "assert finemap['chromosome'].dtype == 'int64'  \n",
    "assert finemap['chromosome'].isin(range(1, 24)).all()  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "337a14ab-c09e-42ae-b737-40f95d2080da",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Adjust `v` in `finemap`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a1c693d8-0ed6-4529-9c3c-3588f5034ce4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Replace 'X' with '23' in 'v' column of finemap\n",
    "finemap['v'] = finemap['v'].str.replace(r'(^X:)', '23:', regex=True)\n",
    "\n",
    "# Assert 'X' is not in 'v' column anymore\n",
    "assert 'X' not in finemap['v']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c30d6716-fce0-4ca7-9f2b-0070bda37dce",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Create `causal` in `gwas`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "1d5eab4c-69e0-4d1e-bbc0-9873645c5141",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "gwas['causal'] = gwas['rsids'].isin(causal['SNP']).astype(int)\n",
    "\n",
    "# Assertions to verify the data manipulation\n",
    "assert 'causal' in gwas.columns  \n",
    "assert gwas['causal'].isin([0, 1]).all()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf557dd6-8479-4055-a68d-ab8d60e9ecfa",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Create `LD` in `gwas`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "38e0d43d-6db4-43a0-be64-ea9463464af8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4143\n"
     ]
    }
   ],
   "source": [
    "# Filter finemap by 'lead_r2' between 0.8 (inclusive) and 1 (exclusive)\n",
    "filtered_finemap = finemap[(finemap['lead_r2'] > 0.8) & (finemap['lead_r2'] < 1)].copy()\n",
    "\n",
    "# Extract the position from 'v' column in filtered_finemap\n",
    "filtered_finemap['position'] = filtered_finemap['v'].str.split(':').str.get(1).astype(int)\n",
    "\n",
    "# Create multi-index for fast selection and sort the index\n",
    "filtered_finemap.set_index(['chromosome', 'position'], inplace=True)\n",
    "filtered_finemap.sort_index(inplace=True)\n",
    "\n",
    "# Check existence in filtered_finemap using index\n",
    "gwas['LD'] = np.asarray(gwas.set_index(['#chrom', 'pos']).index.isin(filtered_finemap.index), dtype=int)\n",
    "\n",
    "ld_sum = gwas['LD'].sum()\n",
    "print(ld_sum)\n",
    "\n",
    "# Assertions to verify the data manipulations\n",
    "assert len(filtered_finemap) <= ld_sum\n",
    "assert '#chrom' in gwas.columns\n",
    "assert 'pos' in gwas.columns\n",
    "assert 'LD' in gwas.columns\n",
    "assert gwas['LD'].isin([0, 1]).all()\n",
    "assert gwas['LD'].sum() >= 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac2fc83d-0075-431d-8e8d-b888c718d056",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Create `lead` in `gwas`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5cb17afd-1596-4fca-8454-86321b4e13d1",
   "metadata": {
    "tags": []
   },
   "source": [
    "`finemap` df:\n",
    "\n",
    "| Column name    | Description                                                        |\n",
    "| -------------- | ------------------------------------------------------------------ |\n",
    "| trait          | endpoint name                                                      |\n",
    "| region         | chr:start-end                                                      |\n",
    "| v              | variant identifier                                                 |\n",
    "| rsid           | rs variant identifier                                              |\n",
    "| chromosome     | chromosome on build GRCh38 (1-23)                                |\n",
    "| position       | position in base pairs on build GRCh38                              |\n",
    "| allele1        | reference allele                                                   |\n",
    "| allele2        | alternative allele (effect allele)                                  |\n",
    "| maf            | minor allele frequency                                             |\n",
    "| beta           | effect size GWAS                                                   |\n",
    "| se             | standard error GWAS                                                |\n",
    "| p              | p-value GWAS                                                       |\n",
    "| mean           | posterior expectation of true effect size                           |\n",
    "| sd             | posterior standard deviation of true effect size                   |\n",
    "| prob           | posterior probability of association                                |\n",
    "| cs             | identifier of 95% credible set (-1 = variant is not part of credible set) |\n",
    "| lead_r2        | r2 value to a lead variant (the one with maximum PIP) in a credible set |\n",
    "| alphax         | posterior inclusion probability for the x-th single effect (x := 1..L where L is the number of single effects (causal variants) specified; default: L = 10) |\n",
    "\n",
    "`gwas` df:\n",
    "\n",
    "| Column name   | Description                                                 |\n",
    "| ------------- | ----------------------------------------------------------- |\n",
    "| #chrom        | chromosome on build GRCh38 (1-23)                           |\n",
    "| pos           | position in base pairs on build GRCh38                       |\n",
    "| ref           | reference allele                                            |\n",
    "| alt           | alternative allele (effect allele)                           |\n",
    "| rsids         | variant identifier                                          |\n",
    "| nearest_genes | nearest gene(s) (comma separated) from variant               |\n",
    "| pval          | p-value from [source]                                        |\n",
    "| mlogp         | -log10(p-value)                                             |\n",
    "| beta          | effect size (log(OR) scale) estimated with [source]          |\n",
    "| sebeta        | standard error of effect size estimated with [source]        |\n",
    "| af_alt        | alternative (effect) allele frequency                        |\n",
    "| af_alt_cases  | alternative (effect) allele frequency among cases            |\n",
    "| af_alt_controls | alternative (effect) allele frequency among controls         |\n",
    "| LD | 1 if variant is in linkage disequilibrium with lead variant, 0 if not        |\n",
    "\n",
    "- create a new column in `gwas named` 'id' combining #chrom, pos, ref, and alt with : as seperator (string)\n",
    "- create a new column in `finemap` named 'id' that is a copy of `v` column (string)\n",
    "- for all target variant (i.e., rows) where `gwas[LD] =1`:\n",
    "    - find the closest lead variant using `chromosome` and `position` in `finemap` for the target variant:\n",
    "        - a lead variant is defined as the SNP with the highest `prob` value within a 15000 base pair distance from the target variant \n",
    "    - if the id for the lead variant exists, add the id of the lead variant to`gwas[lead]` in that row\n",
    "    - else, add \"NaN' string to`gwas[lead]` in that row\n",
    "    - return full `gwas` dataframe "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c18b803a-1a67-4095-bb27-dbb95d88c842",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>rsids</th>\n",
       "      <th>nearest_genes</th>\n",
       "      <th>pval</th>\n",
       "      <th>mlogp</th>\n",
       "      <th>beta</th>\n",
       "      <th>sebeta</th>\n",
       "      <th>af_alt</th>\n",
       "      <th>af_alt_cases</th>\n",
       "      <th>af_alt_controls</th>\n",
       "      <th>causal</th>\n",
       "      <th>LD</th>\n",
       "      <th>id</th>\n",
       "      <th>lead</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>13668</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>rs2691328</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>0.944365</td>\n",
       "      <td>0.024860</td>\n",
       "      <td>-0.005926</td>\n",
       "      <td>0.084918</td>\n",
       "      <td>0.005842</td>\n",
       "      <td>0.005729</td>\n",
       "      <td>0.005863</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1:13668:G:A</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>14773</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>rs878915777</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>0.844305</td>\n",
       "      <td>0.073501</td>\n",
       "      <td>0.010088</td>\n",
       "      <td>0.051369</td>\n",
       "      <td>0.013495</td>\n",
       "      <td>0.013547</td>\n",
       "      <td>0.013485</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1:14773:C:T</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>15585</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>rs533630043</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>0.841908</td>\n",
       "      <td>0.074735</td>\n",
       "      <td>0.031464</td>\n",
       "      <td>0.157751</td>\n",
       "      <td>0.001113</td>\n",
       "      <td>0.001125</td>\n",
       "      <td>0.001110</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1:15585:G:A</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>16549</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>rs1262014613</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>0.343308</td>\n",
       "      <td>0.464316</td>\n",
       "      <td>0.241377</td>\n",
       "      <td>0.254711</td>\n",
       "      <td>0.000561</td>\n",
       "      <td>0.000620</td>\n",
       "      <td>0.000550</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1:16549:T:C</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>16567</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>rs1194064194</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>0.129883</td>\n",
       "      <td>0.886447</td>\n",
       "      <td>0.130736</td>\n",
       "      <td>0.086319</td>\n",
       "      <td>0.004170</td>\n",
       "      <td>0.004250</td>\n",
       "      <td>0.004154</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1:16567:G:C</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20170001</th>\n",
       "      <td>23</td>\n",
       "      <td>155697920</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.027115</td>\n",
       "      <td>1.566790</td>\n",
       "      <td>-0.013475</td>\n",
       "      <td>0.006097</td>\n",
       "      <td>0.290961</td>\n",
       "      <td>0.286054</td>\n",
       "      <td>0.291879</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>23:155697920:G:A</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20170002</th>\n",
       "      <td>23</td>\n",
       "      <td>155698443</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.178417</td>\n",
       "      <td>0.748564</td>\n",
       "      <td>-0.069907</td>\n",
       "      <td>0.051951</td>\n",
       "      <td>0.003259</td>\n",
       "      <td>0.003022</td>\n",
       "      <td>0.003304</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>23:155698443:C:A</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20170003</th>\n",
       "      <td>23</td>\n",
       "      <td>155698490</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.279640</td>\n",
       "      <td>0.553400</td>\n",
       "      <td>-0.020245</td>\n",
       "      <td>0.018725</td>\n",
       "      <td>0.024406</td>\n",
       "      <td>0.024312</td>\n",
       "      <td>0.024423</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>23:155698490:C:T</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20170004</th>\n",
       "      <td>23</td>\n",
       "      <td>155699751</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.078864</td>\n",
       "      <td>1.103120</td>\n",
       "      <td>-0.011284</td>\n",
       "      <td>0.006421</td>\n",
       "      <td>0.244829</td>\n",
       "      <td>0.241257</td>\n",
       "      <td>0.245498</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>23:155699751:C:T</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20170005</th>\n",
       "      <td>23</td>\n",
       "      <td>155700291</td>\n",
       "      <td>CAA</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.071590</td>\n",
       "      <td>1.145150</td>\n",
       "      <td>-0.011563</td>\n",
       "      <td>0.006418</td>\n",
       "      <td>0.245328</td>\n",
       "      <td>0.241724</td>\n",
       "      <td>0.246003</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>23:155700291:CAA:C</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20170006 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          #chrom        pos  ref alt         rsids nearest_genes      pval   \n",
       "0              1      13668    G   A     rs2691328         OR4F5  0.944365  \\\n",
       "1              1      14773    C   T   rs878915777         OR4F5  0.844305   \n",
       "2              1      15585    G   A   rs533630043         OR4F5  0.841908   \n",
       "3              1      16549    T   C  rs1262014613         OR4F5  0.343308   \n",
       "4              1      16567    G   C  rs1194064194         OR4F5  0.129883   \n",
       "...          ...        ...  ...  ..           ...           ...       ...   \n",
       "20170001      23  155697920    G   A           NaN           NaN  0.027115   \n",
       "20170002      23  155698443    C   A           NaN           NaN  0.178417   \n",
       "20170003      23  155698490    C   T           NaN           NaN  0.279640   \n",
       "20170004      23  155699751    C   T           NaN           NaN  0.078864   \n",
       "20170005      23  155700291  CAA   C           NaN           NaN  0.071590   \n",
       "\n",
       "             mlogp      beta    sebeta    af_alt  af_alt_cases   \n",
       "0         0.024860 -0.005926  0.084918  0.005842      0.005729  \\\n",
       "1         0.073501  0.010088  0.051369  0.013495      0.013547   \n",
       "2         0.074735  0.031464  0.157751  0.001113      0.001125   \n",
       "3         0.464316  0.241377  0.254711  0.000561      0.000620   \n",
       "4         0.886447  0.130736  0.086319  0.004170      0.004250   \n",
       "...            ...       ...       ...       ...           ...   \n",
       "20170001  1.566790 -0.013475  0.006097  0.290961      0.286054   \n",
       "20170002  0.748564 -0.069907  0.051951  0.003259      0.003022   \n",
       "20170003  0.553400 -0.020245  0.018725  0.024406      0.024312   \n",
       "20170004  1.103120 -0.011284  0.006421  0.244829      0.241257   \n",
       "20170005  1.145150 -0.011563  0.006418  0.245328      0.241724   \n",
       "\n",
       "          af_alt_controls  causal  LD                  id lead  \n",
       "0                0.005863       0   0         1:13668:G:A  NaN  \n",
       "1                0.013485       0   0         1:14773:C:T  NaN  \n",
       "2                0.001110       0   0         1:15585:G:A  NaN  \n",
       "3                0.000550       0   0         1:16549:T:C  NaN  \n",
       "4                0.004154       0   0         1:16567:G:C  NaN  \n",
       "...                   ...     ...  ..                 ...  ...  \n",
       "20170001         0.291879       0   0    23:155697920:G:A  NaN  \n",
       "20170002         0.003304       0   0    23:155698443:C:A  NaN  \n",
       "20170003         0.024423       0   0    23:155698490:C:T  NaN  \n",
       "20170004         0.245498       0   0    23:155699751:C:T  NaN  \n",
       "20170005         0.246003       0   0  23:155700291:CAA:C  NaN  \n",
       "\n",
       "[20170006 rows x 17 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Make sure that the required columns exist\n",
    "assert 'prob' in finemap.columns\n",
    "assert 'v' in finemap.columns\n",
    "assert 'chromosome' in finemap.columns\n",
    "assert 'position' in finemap.columns\n",
    "\n",
    "assert '#chrom' in gwas.columns\n",
    "assert 'pos' in gwas.columns\n",
    "assert 'ref' in gwas.columns\n",
    "assert 'alt' in gwas.columns\n",
    "assert 'LD' in gwas.columns\n",
    "\n",
    "# Create a new column in gwas named 'id' combining #chrom, pos, ref, and alt\n",
    "gwas['id'] = gwas['#chrom'].astype(str) + ':' + gwas['pos'].astype(str) + ':' + gwas['ref'] + ':' + gwas['alt']\n",
    "\n",
    "# Create a new column in finemap named 'id' from finemap['v']\n",
    "finemap['id'] = finemap['v']\n",
    "\n",
    "# Find the lead variants in finemap\n",
    "lead_variants = finemap[finemap['lead_r2'] == 1].copy()\n",
    "\n",
    "def get_highest_prob_lead(row):\n",
    "    # Filter lead_variants for the same chromosome\n",
    "    same_chrom = lead_variants[lead_variants['chromosome'] == row['#chrom']]\n",
    "    \n",
    "    # If there are no lead variants on the same chromosome, return 'NaN'\n",
    "    if same_chrom.empty:\n",
    "        return 'NaN'\n",
    "\n",
    "    # Filter variants within a 15000 base pair distance from the target variant\n",
    "    same_chrom = same_chrom[abs(same_chrom['position'] - row['pos']) <= 7500]\n",
    "\n",
    "    # If there are no lead variants within the distance, return 'NaN'\n",
    "    if same_chrom.empty:\n",
    "        return 'NaN'\n",
    "\n",
    "    # Sort same_chrom by 'prob' in descending order and select the first row\n",
    "    highest_prob = same_chrom.sort_values(by='prob', ascending=False).iloc[0]\n",
    "\n",
    "    return highest_prob['id']\n",
    "\n",
    "# Initialize 'lead' column with 'NaN'\n",
    "gwas['lead'] = 'NaN'\n",
    "\n",
    "# Apply the function to each row in gwas where LD = 1\n",
    "for index, row in gwas[gwas['LD'] == 1].iterrows():\n",
    "    gwas.at[index, 'lead'] = get_highest_prob_lead(row)\n",
    "\n",
    "# Return full gwas dataframe \n",
    "gwas\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "e43cf595-fa9f-47e5-a0a9-38d00cf78f55",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of non-'NaN' values in 'lead' column:  1125\n"
     ]
    }
   ],
   "source": [
    "# Count non-\"NaN\" values in 'lead' column\n",
    "non_na_count = len(gwas[gwas['lead'] != 'NaN'])\n",
    "\n",
    "# Print the count\n",
    "print(\"Number of non-'NaN' values in 'lead' column: \", non_na_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5e503e8-1146-47d2-a25c-0f57905d4994",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Extract `trait` from `finemap` to `gwas`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "159fbc16-3d9c-4e80-b8dd-271a2088b20d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "unique_trait = finemap['trait'].unique()\n",
    "trait_string = unique_trait[0]\n",
    "gwas['trait'] = trait_string"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a4694be-acb8-4d22-9053-d985fe51fef6",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Export csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ca8ebd33-8d99-4a37-85f0-c707d04333b7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#gwas.to_csv('gwas-causal.csv', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
